<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies).

In [ ]:
# текст запроса
query_3_1 = f'''

select
    count(id)

from
    vacancies
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
print(f'Количество вакансий в базе: {df.iloc[0,0]}')

Количество вакансий в базе: 49197


<ipython-input-36-f54352845692>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


2. Напишите запрос, который посчитает количество работодателей (таблица employers).

In [ ]:
# текст запроса
query_3_2 = f'''

select
    count(id)

from
    employers
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
print(f'Количество работадателей: {df.iloc[0,0]}')

Количество работадателей: 23501


<ipython-input-34-f65d420aa9ee>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [ ]:
# текст запроса
query_3_3 = f'''

select
    count(id)

from
    areas
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
print(f'Количество регионов: {df.iloc[0,0]}')

Количество регионов: 1362


<ipython-input-32-83d3de90dc1f>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [ ]:
# текст запроса
query_3_4 = f'''

select
    count(id)

from
    industries
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
print(f'Количество сфер деятельности: {df.iloc[0,0]}')

Количество сфер деятельности: 294


<ipython-input-30-021a24d1b3fc>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


***

**Выводы по предварительному анализу данных:**
1. На одного условного работадателя приходится чуть более 2 вакансий в среднем.
2. Примерно 17 работадателей в среднем будет приходится на регион.
3. На одну сферу деятельности приходится в среднем примерно 167 вакансий.
4. На одну сферу деятельности приходится примерно 140 работадетелей в среднем.
5. Среднее количество вакансий на регион составит примерно 36.


# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [ ]:
# текст запроса
query_4_1 = f'''

SELECT
    a.name,
    count(v.id) cnt

FROM
    vacancies v
    LEFT JOIN areas a ON v.area_id = a.id

GROUP BY
    a.name

ORDER BY
    cnt DESC

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
display(df.head(5))

<ipython-input-58-e90e1fc19590>:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
# текст запроса
query_4_2 = f'''

SELECT
    COUNT(*)

FROM
    vacancies

WHERE
    (salary_from IS NOT NULL AND salary_to IS NOT NULL) OR (salary_from IS NOT NULL AND salary_to IS NULL) OR (salary_from IS NULL AND salary_to IS NOT NULL)

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
print(f'Количество вакансий у которых заполнена хотябы одна строка поля по зарплате: {df.iloc[0,0]}')

Количество вакансий у которых заполнена хотябы одна строка поля по зарплате: 24073


<ipython-input-12-75163b26f36c>:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [ ]:
# текст запроса
query_4_3 = f'''

SELECT
    AVG(salary_from) lower,
    AVG(salary_to) upper

FROM
    vacancies

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
print(f'Среднее значение нижнего предела по зарплате: {round(df.iloc[0,0])}')
print(f'Среднее значение верхнего предела по зарплате: {round(df.iloc[0,1])}')

Среднее значение нижнего предела по зарплате: 71065
Среднее значение верхнего предела по зарплате: 110537


<ipython-input-36-35e090b72cbf>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
# текст запроса
query_4_4 = f'''

SELECT
    schedule,
    employment,
    COUNT(*)

FROM
    vacancies

GROUP BY
    schedule, employment

ORDER BY
    COUNT(*) DESC

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
display(df.iloc[1,0:2])

<ipython-input-60-5faa76cae9a5>:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



schedule      Удаленная работа
employment    Полная занятость
Name: 1, dtype: object

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [ ]:
# текст запроса
query_4_5 = f'''

SELECT
    experience,
    COUNT(*)

FROM
    vacancies

GROUP BY
    experience

ORDER BY
    COUNT(*)

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
display(df)

<ipython-input-114-9bd57a49065f>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

**Выводы по детальному анализу вакансий:**
1. У чуть более половины всех вакансий не указаны зарплатные ожидания.
2. Примерно каждая одинадцатая вакансия из Москвы.
3. Из-за специфики соискателей довольно большой популярностью пользуется удаленная работа.
4. Наибольшим спросам на рынке труда в ИТ пользуются люди у которых стаж работы от 1 года до 3 лет трудовой деятельности.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [80]:
# текст запроса
query_5_1 = f'''

SELECT
    e.name,
    count(v.id)

FROM
    vacancies v
    LEFT JOIN employers e ON v.employer_id = e.id

GROUP BY
    e.name

ORDER BY
    count(*) DESC

LIMIT 5

'''

In [81]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
print(f'Первое место по количеству вакансий занимает: {df.iloc[0,0]}')
print(f'Пятое место по количеству вакансий занимает: {df.iloc[4,0]}')


Первое место по количеству вакансий занимает: Яндекс
Пятое место по количеству вакансий занимает: Газпром нефть


<ipython-input-81-d664596ab463>:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
# текст запроса
query_5_2 = f'''

SELECT
    a.name as area,
    COUNT(v.id) as cnt_vac,
    COUNT(e.id) as cnt_emp

FROM
    areas a
    LEFT JOIN employers e ON  a.id = e.area
    LEFT JOIN vacancies v ON  a.id = v.area_id

WHERE
    v.id IS NULL

GROUP BY
    a.name

ORDER BY
     cnt_emp DESC

LIMIT 1

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
print(f'Название региона в котором наибольшее количество работодателей при отсутствии вакансий: {df.iloc[0,0]}')

Название региона, в котором наибольшее количество работодателей, при отсутствии вакансий: Россия


<ipython-input-56-974be7780211>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [ ]:
# текст запроса
query_5_3 = f'''

SELECT
    employer_id,
    COUNT(DISTINCT area_id)

FROM
    VACANCIES

GROUP BY
    employer_id

ORDER BY
      COUNT(DISTINCT area_id) DESC

LIMIT 1

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
print(f'Максимальное количество регионов, в которых работодатель публикует свои вакансии: {df.iloc[0,1]}')

<ipython-input-81-ee3f220a5ddf>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


Максимальное количество регионов, в которых работодатель публикует свои вакансии: 181


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [ ]:
# текст запроса
query_5_4 = f'''

SELECT
    COUNT(*)

FROM
    employers e
    LEFT JOIN employers_industries ei ON e.id = ei.employer_id
    LEFT JOIN industries i ON ei.industry_id = i.id

WHERE
    i.name IS NULL

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
print(f'Количество работодателей, у которых не указана сфера деятельности : {df.iloc[0,0]}')

Количество работодателей, у которых не указана сфера деятельности : 8419


<ipython-input-115-7af673519131>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [ ]:
# текст запроса
query_5_5 = f'''

SELECT
    e.name company_name,
    COUNT(DISTINCT i.name) cnt_ind

FROM
    employers e
    LEFT JOIN employers_industries ei ON e.id = ei.employer_id
    LEFT JOIN industries i ON ei.industry_id = i.id

GROUP BY
    e.name

HAVING
    COUNT(DISTINCT i.name) = 4

ORDER BY
    company_name

OFFSET 2

LIMIT 1

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
print(f'Название компании, находящейся на третьем месте, у которых указано четыре сферы деятельности : {df.iloc[0,0]}')

<ipython-input-160-8f97ca80d189>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


Название компании, находящейся на третьем месте, у которых указано четыре сферы деятельности : 2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [ ]:
# текст запроса
query_5_6 = f'''

SELECT
    COUNT(*) cnt_employers

FROM
    employers e
    LEFT JOIN employers_industries ei ON e.id = ei.employer_id
    LEFT JOIN industries i ON ei.industry_id = i.id

WHERE
    i.name = 'Разработка программного обеспечения'

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
print(f'Количество работодателей, в качестве сферы деятельности которых указана "Разработка программного обеспечения": {df.iloc[0,0]}')

Количество работодателей, в качестве сферы деятельности которых указана "Разработка программного обеспечения": 3553


<ipython-input-169-d6a517222ebd>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8).

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API.

In [ ]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup # Импортируем библиотеку requests
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Определяем адрес страницы
response = requests.get(url)  # Выполняем GET-запрос
page = BeautifulSoup(response.text, 'html.parser')
links = page.find_all('a')
cities = tuple([link.text for link in links[38:54]])
print(cities)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [ ]:
# текст запроса
query_5_7 = f'''

SELECT
    a.name,
    COUNT(v.id) cnt

FROM
    areas a
    JOIN vacancies v ON a.id = v.area_id
    JOIN employers e ON e.id = v.employer_id

WHERE
    e.name = 'Яндекс' and a.name in {cities}

GROUP BY
    a.name

UNION ALL

SELECT
         'Total',
         SUM(1)

FROM
    areas a
    JOIN vacancies v ON a.id = v.area_id
    JOIN employers e ON e.id = v.employer_id

WHERE
    e.name = 'Яндекс' and a.name in {cities}

ORDER BY
    cnt

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
display(df)

<ipython-input-60-7619c91123bd>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

**Выводы по анализу работодателей:**
1.   Всего "Яндекс" предоставляет 1933 вакансии, что является примерно 4% от общего количества всех вакансий.
2.   У "Газпром Нефти" количество вакансий почти в 6 раз меньше, чем у "Яндекса".
3.   Известно, что у 8419 работодателей не указана сфера деятельности, что является 36% от общего числа всех работадателей.
4.   Примерно у одной из 20 компаний указаны 4 сферы деятельности.
5.   Примерно у 15% работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».
6.   Примерно каждая сотая вакансия в Москве принадлежит "Яндексу".
7.   Одна из четырех вакансий "Яндекса" размещена в городах миллионниках.
8.   Примерно 11 вакансий из 100 "Яндекса" для Москвы.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [ ]:
# текст запроса
query_6_1 = f'''

SELECT
    COUNT(id)

FROM
    vacancies

WHERE
    (name LIKE '%data%') OR (name LIKE '%данн%') OR (name LIKE 'data%') OR (name LIKE 'данн%') OR
    (name LIKE '%Data%') OR (name LIKE '%Данн%') OR (name LIKE 'Data%') OR (name LIKE 'Данн%') OR
    (name LIKE '%DATA%') OR (name LIKE '%ДАНН%') OR (name LIKE 'DATA%') OR (name LIKE 'ДАНН%')

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
print(f'Количество вакансий, которые имеют отношение к данным : {df.iloc[0,0]}')

Количество вакансий, которые имеют отношение к данным : 1771


<ipython-input-154-e49e098a405b>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста?
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.


In [ ]:
# текст запроса
query_6_2 = f'''

SELECT
    COUNT(id)

FROM
    vacancies

WHERE
    ( experience = 'Нет опыта' OR employment = 'Стажировка' OR name ILIKE '%junior%' )
    AND
    ( name ILIKE '%data scientist%' OR
      name ILIKE '%data science%' OR
      name ILIKE '%исследователь данных%' OR
      name ILIKE '%machine learning%' OR
      name ILIKE '%машинн%обучен%' OR
      name ILIKE '%ML%' AND name NOT ILIKE '%HTML%' )

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
print(f'Количество вакансий по указанному выше условию составляет : {df.iloc[0,0]}')

<ipython-input-44-43e0a2245c9a>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


Количество вакансий по указанному выше условию составляет : 51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# текст запроса
query_6_3 = f'''

SELECT
    COUNT(id)

FROM
    vacancies

WHERE
    ( name ILIKE '%data scientist%' OR
      name ILIKE '%data science%' OR
      name ILIKE '%исследователь данных%' OR
      name ILIKE '%machine learning%' OR
      name ILIKE '%машинн%обучен%' OR
      name ILIKE '%ML%' AND name NOT ILIKE '%HTML%' )
    AND
    ( key_skills ILIKE '%SQL%' OR key_skills ILIKE '%postgres%' )

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
print(f'Количество вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres : {df.iloc[0,0]}')

<ipython-input-37-e2e63f9bde52>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


Количество вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres : 229


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# текст запроса
query_6_4 = f'''

SELECT
    COUNT(id)

FROM
    vacancies

WHERE
    ( name ILIKE '%data scientist%' OR
      name ILIKE '%data science%' OR
      name ILIKE '%исследователь данных%' OR
      name ILIKE '%machine learning%' OR
      name ILIKE '%машинн%обучен%' OR
      name ILIKE '%ML%' AND name NOT ILIKE '%HTML%' )
    AND
      (key_skills ILIKE '%Python%')

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
print(f'Количество вакансий, в которых в качестве ключевого навыка указан Python : {df.iloc[0,0]}')

<ipython-input-107-bc88d3cf876f>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


Количество вакансий, в которых в качестве ключевого навыка указан Python : 357


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# текст запроса
query_6_5 = f'''

SELECT
    round((AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1)), 2)

FROM
    vacancies

WHERE
    ( name ILIKE '%data scientist%' OR
      name ILIKE '%data science%' OR
      name ILIKE '%исследователь данных%' OR
      name ILIKE '%machine learning%' OR
      name ILIKE '%машинн%обучен%' OR
      name LIKE '%ML%' AND name NOT ILIKE '%HTML%' )
      AND (key_skills IS NOT NULL)

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
print(f'Количество ключевых навыков в среднем для DS вакансий : {df.iloc[0,0]}')

<ipython-input-133-f7a7b376e5d1>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


 Количество ключевых навыков в среднем для DS вакансий : 6.41


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*).

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа.

In [ ]:
# текст запроса
query_6_6 = f'''

SELECT
    experience,
    round(AVG(coalesce((salary_from + salary_to) / 2, salary_from, salary_to)))

FROM
    vacancies

WHERE
    ( name ILIKE '%data scientist%' OR
      name ILIKE '%data science%' OR
      name ILIKE '%исследователь данных%' OR
      name ILIKE '%machine learning%' OR
      name ILIKE '%машинн%обучен%' OR
      name LIKE '%ML%' AND name NOT ILIKE '%HTML%' )
      AND
    ( (salary_from IS NOT NULL AND salary_to IS NOT NULL) OR
      (salary_from IS NULL AND salary_to IS NOT NULL) OR
      (salary_from IS NOT NULL AND salary_to IS NULL) )

GROUP BY
    experience

'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
print(f'Средняя зарплата дата-сайентиста с опытом работы от 3 до 6 лет : {df.iloc[2,1]}')

<ipython-input-52-d54466ee65ec>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


Средняя зарплата дата-сайентиста с опытом работы от 3 до 6 лет : 243115.0


***

**Выводы по предметному анализу:**
1. Лишь 3.6% от общего числа всех вакансий имеют отношение к данным или примерно одна вакансия на каждые 28 других.
2. Количество вакансий для начинающих дата-саентистов составляет около 3% от всех вакансий для DS.
3. Вакансии, в которых ключевым навыком будет PYTON встречается в полтора раза чаще, чем если бы это
 был ключевой навык SQL и postgres для DS вакансий.
4. Примерно каждая пятая вакансия для DS будет с ключевым навыком PYTON.
5. Средняя зарплата, которую предлагают для тех, кто работает в Data Science от 1 года до 3 лет выше, чем у новичков в этой сфере почти в 2 раза, а для тех кто работает от 3 до 6 лет - в три раза.

# **Проведем дополнительные ислледования и сделаем выводы.**

In [9]:
import plotly # Подключим графическую библиотеку plotly express i pandas
import plotly.express as px
import pandas as pd

In [6]:
# текст запроса
query_7_1 = f'''

SELECT
    experience,
    count(id)

FROM
    vacancies

GROUP BY
    experience

ORDER BY
    count(id) DESC

'''

In [9]:
# результат запроса
df = pd.read_sql_query(query_7_1, connection)
fig = px.pie(df, values='count', names='experience', height= 500, width= 700, title='Круговая диаграмма вакансий по уровню опыта работы')
fig.show()

<ipython-input-9-d2e70cc468d8>:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [54]:
# текст запроса
query_7_2 = f'''

SELECT
    experience,
    round(AVG(coalesce((salary_from + salary_to) / 2, salary_from, salary_to))) as salary_avg

FROM
    vacancies

WHERE
    (salary_from IS NOT NULL AND salary_to IS NOT NULL) OR
    (salary_from IS NULL AND salary_to IS NOT NULL) OR
    (salary_from IS NOT NULL AND salary_to IS NULL)

GROUP BY
    experience

ORDER BY
    salary_avg DESC

'''

In [57]:
# результат запроса
df = pd.read_sql_query(query_7_2, connection)
display(df)

<ipython-input-57-eeab80e0ba75>:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience  salary_avg
0         Более 6 лет    166256.0
1       От 3 до 6 лет    133211.0
2  От 1 года до 3 лет     76541.0
3           Нет опыта     40472.0

In [58]:
# текст запроса
query_7_3 = f'''

SELECT
    employment,
    count(id)

FROM
    vacancies

GROUP BY
    employment

'''

In [59]:
# результат запроса
df = pd.read_sql_query(query_7_3, connection)
fig = px.pie(df, values='count', names='employment', height= 500, width= 700, title='Круговая диаграмма вакансий по уровню занятости')
fig.show()

<ipython-input-59-e2521063911d>:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [60]:
# текст запроса
query_7_4 = f'''

SELECT
    schedule,
    count(id)

FROM
    vacancies

GROUP BY
    schedule

'''

In [62]:
# результат запроса
df = pd.read_sql_query(query_7_4, connection)
fig = px.pie(df, values='count', names='schedule', height= 500, width= 700, title='Круговая диаграмма соотношения вакансий от вида графика работы')
fig.show()

<ipython-input-62-45d2c3dbe1a1>:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [5]:
# текст запроса
query_7_5 = f'''

SELECT
    COUNT(v.id)

FROM

    vacancies v
    LEFT JOIN employers_industries ei ON v.employer_id = ei.employer_id
    LEFT JOIN industries i ON ei.industry_id = i.id

WHERE
    i.name = 'Разработка программного обеспечения'

'''

In [6]:
# результат запроса
df = pd.read_sql_query(query_7_5, connection)
print(f'Количество вакансий у которых в сфере деятельности указано "Разработка программного обеспечения" : {df.iloc[0,0]}')

<ipython-input-6-ee714cc7b5b1>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_5, connection)


Количество вакансий у которых в сфере деятельности указано "Разработка программного обеспечения" : 12499


In [8]:
# текст запроса
query_7_6 = f'''

SELECT
    a.name,
    count(a.name) cnt

FROM
    vacancies v
    LEFT JOIN areas a ON v.area_id = a.id

 WHERE
    (v.name ILIKE '%data scientist%') OR
    (v.name ILIKE '%data science%') OR
    (v.name ILIKE '%исследователь данных%') OR
    (v.name ILIKE '%machine learning%') OR
    (v.name ILIKE '%машинн%обучен%') OR
    ( (v.name ILIKE '%ML%') AND (v.name NOT ILIKE '%HTML%') )

GROUP BY
    a.name

ORDER BY
    cnt DESC

LIMIT 5

'''

In [9]:
# результат запроса
df = pd.read_sql_query(query_7_6, connection)
display(df)# выводим на экран топ-5 городов по количеству data science вакансий

<ipython-input-9-2dda8c2b428f>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_6, connection)


name  cnt
0           Москва  225
1  Санкт-Петербург   69
2      Новосибирск   23
3           Алматы   22
4  Нижний Новгород   20

In [10]:
# текст запроса
query_7_7 = f'''

SELECT
    e.name,
    count(e.name) cnt

FROM
    vacancies v
    LEFT JOIN employers e ON e.id = v.employer_id

 WHERE
    (v.name ILIKE '%data scientist%') OR
    (v.name ILIKE '%data science%') OR
    (v.name ILIKE '%исследователь данных%') OR
    (v.name ILIKE '%machine learning%') OR
    (v.name ILIKE '%машинн%обучен%') OR
    ( (v.name ILIKE '%ML%') AND (v.name NOT ILIKE '%HTML%') )

GROUP BY
    e.name

ORDER BY
    cnt DESC

LIMIT 5

'''

In [11]:
# результат запроса
df = pd.read_sql_query(query_7_7, connection)
# выводим на экран топ-5 компаний по количеству data science вакансий
display(df)

<ipython-input-11-707bfd6138a6>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_7, connection)


name  cnt
0                   СБЕР   39
1        Bell Integrator   25
2         Банк ВТБ (ПАО)   18
3                     VK   15
4  Positive Technologies   11

# Общие выводы по проекту:
1.  Как видно из диаграммы ниже, чуть более половины всех вакансий (53%) от работадателей ищут работников от с опытом от года до 3 лет. На мой взгляд это может быть связано с тем, что работадателей устраивает соотношение и зарплатных ожиданий и компетенции данной категории соискателей. Выведим на экран таблицу с зарплатами для вакансий в зависимости от опыта. На ней четко видна прямая зависимость средней зарплаты от увеличения опыта работы, что подтвержает сделанное предположение. Можно привести дополнительное исследование в подтвеждение этого предположения, но здесь я этого делать не буду.
2.  Из круговой диаграммы вакансий по уровню занятости ниже (запрос 7_3) видно, что практичести все вакансии (93.6%) относятся к полной занятости соискателей.
3.  Из круговой диаграммы вакансий по типу рабочего графика ниже (запрос 7_4) видно, что абсолютное большинство вакансии (74%) относятся к полному рабочему дню соискателей.
4.  Из запроса 7_5 известно, что количество вакансий, у которых в сфере деятельности указано "Разработка программного обеспечения" составляет 12499. Известно,что количество работадателей из этой же сферы : 3553. Следовательно в среднем на одного работадателя должно приходиться 3,5 вакансии
5.  Из запроса 7_6 известно, что количество data_science вакансий для Москвы : 225, а из запроса 4_1 известно, что общее количество всех вакансий для Москвы : 5333. Следовательно примерно 4 из 100 вакансий будут для data_science.
6.  Из запроса 7_7 извесно, что СБЕР предлагает 39 data science вакансий. Из запроса 5_1 четвертое место занимает СБЕР с общим числом вакансий 428. Значит у СБЕРа около 9% всех вакансий - это data_science вакансии.

In [ ]:
connection.close()